memo.js

```diff
@@ -0,0 +1,30 @@
+NO_VALUE = Symbol("NO_VALUE");
+
+class Trie {
+  constructor(value) {
+    this.value = value;
+    this.children = new Map();
+  }
+
+  getOrInsert(path, fn, args) {
+    if (path.length === 0) {
+      if (this.value === NO_VALUE) {
+        this.value = fn(...args);
+      }
+      return this.value;
+    }
+    const [head, ...tail] = path;
+    if (!this.children.has(head)) {
+      this.children.set(head, new Trie(NO_VALUE));
+    }
+    return this.children.get(head).getOrInsert(tail, fn, args);
+  }
+}
+
+const memoize = (fn) => {
+  const memory = new Trie(NO_VALUE);
+
+  return (...args) => memory.getOrInsert(args, fn, args);
```

Comment:
I'm confused by `...args`. Where are those coming from?

Reply:
Much like this Ruby syntax:

```ruby
def blorp(*args)
  args.each_with_index { |e, i| puts "##{i}: #{e}" }
end
blorp("hello", "nice", "world")
#=> #0: hello
#=> #1: nice
#=> #2: world
```

The `...args` syntax in ES collects all of the function's arguments into an array.

```es6
const blorp = (...args) => { args.forEach((e, i) => { console.log(`#${i}: ${e}`) }) }
blorp("hello", "nice", "world")
//=> #0: hello
//=> #1: nice
//=> #2: world
```

Reply:
Hmm ok but in the context of the `memoize` constant, where do they come from? The `fn` on line 24?